R/slice.R

```diff
@@ -48,16 +48,18 @@ vec_slice.default <- function(x, i) {
   d <- vec_dims(x)
   if (d == 1) {
     if (is.object(x)) {
-      x[i]
+      out <- x[i]
     } else {
-      x[i, drop = FALSE]
+      out <- x[i, drop = FALSE]
     }
   } else if (d == 2) {
-    x[i, , drop = FALSE]
+    out <- x[i, , drop = FALSE]
   } else {
     miss_args <- rep(list(missing_arg()), d - 1)
-    eval_bare(expr(x[i, !!!miss_args, drop = FALSE]))
+    out <- eval_bare(expr(x[i, !!!miss_args, drop = FALSE]))
   }
+
+  vec_restore(out, x)
```

Comment:
It seems weird to restore if we've called `[` on something that's an S3 object?

Reply:
Good point, I'll change to only call it on non-OO objects with dims.

Reply:
Actually, calling `vec_restore()` here is congruent with how `vec_slice()` automatically restores the class and attributes. The `[.foo` methods should call `vec_slice()` but not `vec_restore()`. Then the call path for a foo object with dims is `[` -> `vec_slice()` -> `unclass()` -> `[` -> `vec_restore()`.

I have simplified `vec_slice_fallback()` (which is only called with shaped objects and should be removed in the future when we support those natively) and added a missing `unclass()`.

Still need to figure out what to do when additional arguments are passed to `[` methods. Currently higher dimensions are ignored.

Reply:
I guess the current situation is confusing because of contradictory use cases:

- Using `vec_slice()` from our `[` methods for consistency of behaviour.
- Calling `[` from `vec_slice()` to handle shaped objects because we don't have native slicing for these in vctrs.
- Calling `[` from `vec_slice()` to handle S3 objects as a fallback.

With `[.vctrs_vctr` we'd like to restore after slicing shaped objects with `[`. With foreign classes we don't want to restore.

This will be less confusing once we slice shape natively. I'll try to find an interim solution.

Reply:
You can leave it for now and tackle in a future PR if you want.